### PR TITLE
Process TracWiki links and formatting. (#4)

### PR DESCRIPTION
**Pull request by danuker**
Scope
=====

There are TracWiki markup links which need to be converted. Here are examples of files (taken from https://github.com/chevah/trac-to-github/issues/2#issuecomment-764700864_ ):

* TracWiki markup - https://trac.chevah.com/wiki/WikiStart?action=edit
* RST markup with Trac directives - https://trac.chevah.com/wiki/Development/Python?action=edit
* Mixed of RST with TracWiki macros (see bottom) - https://trac.chevah.com/wiki/Administrative?action=edit

At first, we can use this ticket to convert RST Trac links directives.

Later we should look at converting whole TracWiki pages.

Changes
=======

TracWiki links should now work. I have added a new match case to `_trac_rst_wiki_to_github_links`, as well as new methods that handle TracWiki link syntax (with or without link text different from the article).

Also Trac tickets should work. There were only 3 of them, but it was easy enough to automate. 

Only one malformed `:trac:` reference remains, which will be addressed manually.

TracWiki syntax: the script handles the following:
* title/heading syntax: heading (===) and subheading (---)
* lists: it unindents single-space-quoted lists, and adds an empty line before the first list element. The majority of files looks good; but nesting conversion is not proper.

What is not done
---------

About lists: I did not properly implement nesting. I started to, but it turned out to be too difficult for me, considering the diminishing returns. An exhaustive (as of 2021-01-27) list of failures:
* Architecture/Agent/General section "Remote file", under "steps"
* Development/GetToKnow section "Products documentation and download pages".
* Infrastructure/OutOfOrderMachines Mosna under "network cards"
* Infrastructure/Services/VPN under "Overton"
* Requirements/Vision, last two list items on page
* Support/General section "Initial Assessment"
* Support/CustomerDataBase/Template.rst and some files created from that template.

In addition, `{{{ }}}` code blocks need to be converted manually, by searching for `{{{` in a commit before stripping the `{{{#!rst }}}` armor.

How to try and test the changes
===============================

reviewers: @adiroiban 

First, copy the `config.py.sample` to `config.py`, because the Trac URL is specified as a config option.

Then, continue testing like PR #3:

> Run tests using `unittest` from the project directory:
> `python3.8 -m unittest`
> ## Checking output
> 
> I did not find an easy way to verify the GitHub link behavior. The only way to properly test is to run the conversion, then push and check the wiki.
> 
> There are some things you can still check locally:
> 
>   * run the script, and then look at the diff using gitk, with the "Ignore space change" checkbox checked.
> 
>   * check what `:trac:` markers are left

### CODE REVIEW

**Comment by danuker** [comment]
This is ready for review.

I have some questions:
- Is there any way to further deduplicate all the `_trac[...]` methods? They seem to rhyme to me, but could not quite extract another method from sections that repeat.
- Are the tiny methods overkill? (`_underline`, `_matches`, `_sub`)? To me they look okay.

**Comment by danuker** [comment]
needs-review

File: test/test_wiki_trac_rst_convert.py [hunk 1]
```diff
@@ -107,6 +107,168 @@ def test_several_trac_wiki_rst_links_with_content(self):
...
+        self.assertConvertedContent(
+            'Subheading\n----------\n',
+            '== Subheading =='
+        )
+        self.assertConvertedContent(
+            'Subheading\n----------\n\n'
+            'Some text\n',
+
+            '== Subheading ==\n\n'
+            'Some text'
```
Comment (adiroiban) [inline test/test_wiki_trac_rst_convert.py:202]: Just a minor test that is missing.

Conversion from source

```
== Subheading ==
Test right below the heading
```


It would be nice to convert it as something like this with one extra newline added...

```
Subheading
----------

Test right below the heading
```

but this conversion is also ok


```
Subheading
----------
Test right below the heading
```

no need to spend more time here

File: wiki_trac_rst_convert.py [hunk 2]
```diff
@@ -38,42 +40,190 @@ def convert_content(text: str):
...
+            text
+        )
+    return text
+
+
+def _tracwiki_heading_to_rst_heading(text: str):
+    """
+    Convert TracWiki headings (titles surrounded by "=") to RST (titles
+    underlined with "=").
+    """
```
Comment (adiroiban) [inline wiki_trac_rst_convert.py:146]: Maybe use some simple example in the docstring to explain how the conversion is done.

No need for all the details... we can have corner cases explained in the test code.

```suggestion
def _tracwiki_1st_heading_to_rst_heading(text: str):
    """
    Convert TracWiki  1st level headings to RST heading.
    
    TracWiki:
    
    = Some Top Heading =
    Content here
    
    RST conversion:
    
    Some Top Heading
    ----------------
    Content here
    """
```

File: test/test_wiki_trac_rst_convert.py [hunk 3]
```diff
@@ -3,10 +3,10 @@
 from wiki_trac_rst_convert import convert_content
 
 
-class TracRstToVanillaRst(unittest.TestCase):
+class TracToGithubRst(unittest.TestCase):
```
Comment (adiroiban) [inline test/test_wiki_trac_rst_convert.py:6]: naming is complicated  ... but I think that the correct speling is GitHub ...also, when we have an abreviation like RST or URL, we keep upper case in the Python name.

so we prefer `getURL`  and not `getUrl` 

```suggestion
class TracToGitHubRST(unittest.TestCase):
```

**Review by adiroiban — APPROVED** [review]
Well done.


--------------

We also have the 3rd level of heading ... but I think that this is only in a few pages

8 matches across 8 files

But we need to create a list in README or a script (python or shell) to identify the pages and things that will need manual conversion.

-------------

I would like you to continue working on this and have a PR in which the `[[PageOutline]]` macro is fixed and in which every page will have a local table of content inserted.

And then one final PR in which `[[TitleIndex(Development/)]]`macro is fixed in the footer... and we should be ready with the conversion.

-------

I only left minor comments. Feel free to addess as you wish and merge this.

For this one time conversion code I have not looks into details on how the code is written .
I was focused on the tests and they are great :)

--------

The GitHub action integration is more about having you familiarized with GitHub action continuous testing infrastructure ... as we use it a lot.

We still have Buildbot, but the plan is to migrate to GitHub actions. 

Thanks!

File: test/test_wiki_trac_rst_convert.py [hunk 4]
```diff
@@ -107,6 +107,168 @@ def test_several_trac_wiki_rst_links_with_content(self):
...
+            " * item 2\n"
+            " * item 3\n"
+        )
+
+    def test_list_after_paragraph(self):
+        """
+        Separates lists from paragraphs by one empty line.
+
+        In RST, when lists follow a paragraph without an empty line
+        inbetween, they fail to parse as lists.
```
[the remark this answers is not in the document]
Comment (danuker) [reply test/test_wiki_trac_rst_convert.py:227]: I will leave it as is (add an empty line between), because GitHub does some very strange things with the syntax. 

It either:
* (not pictured) does not parse it as a list, if no newline nor one-space indent before list items, or
* it makes the preceding sentence bold and italic:

gh-rst

File: test/test_wiki_trac_rst_convert.py (hunk 1 above)
Comment (danuker) [reply test/test_wiki_trac_rst_convert.py:202]: The test right below this one (with "Some text") ensures two `\n`s already. But I will format it like [you suggested for the Heading test](https://github.com/chevah/trac-to-github/pull/6#discussion_r565509991), since it is clear that the `\n`s on the same line are confusing :)

I wondered what `black` does. It's a problem, but they are [aware of it](https://github.com/psf/black/issues/330). Look what it made of the two tests:

```python
self.assertConvertedContent("Subheading\n" "----------", "== Subheading ==")
self.assertConvertedContent(
    "Subheading\n" "----------\n" "\n" "Some text",
    "== Subheading ==\n" "\n" "Some text",
)
```
